### PR TITLE
ScpSl update

### PR DESCRIPTION
**Pull request by IRacle1**
Fix for many things, at least exiled starts normally without errors

### CODE REVIEW

File: Exiled.Events/Patches/Events/Player/ChangingRole.cs [hunk 1]
```diff
@@ -30,7 +30,7 @@ namespace Exiled.Events.Patches.Events.Player
     using Player = Handlers.Player;
 
     /// <summary>
-    ///     Patches <see cref="PlayerRoleManager.InitializeNewRole(RoleTypeId, RoleChangeReason, Mirror.NetworkReader)" />
+    ///     Patches <see cref="PlayerRoleManager.InitializeNewRole(RoleTypeId, RoleChangeReason, RoleSpawnFlags, Mirror.NetworkReader)" />
     ///     .
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/ChangingRole.cs:34]:
```suggestion
```

File: Exiled.Events/Patches/Events/Player/FlippingCoin.cs [hunk 2]
```diff
@@ -39,19 +39,7 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
-
-            // Remove both NW events and logic
-            newInstructions.RemoveRange(index, 45);
-
-            // Redirect the old labels to the new starting point
-            newInstructions[index].WithLabels(oldLabels);
-
-            offset = -1;
-            index = newInstructions.FindLastIndex(instruction => instruction.LoadsField(Field(typeof(Coin), nameof(Coin.RateLimiter)))) + offset;
+            int index = newInstructions.FindIndex(instruction => instruction.LoadsConstant(109)) + offset;
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/FlippingCoin.cs:42]: If the flip gets prevented from the NwAPI `PlayerPreCoinFlip` event, the Exiled event is never called.
We could either suppress their event (not ideal) or remove the return from their switch and add their `CoinFlipCancellation` enum to our event, to handle its logic by ourselves.

**Review by iopietro — CHANGES_REQUESTED** [review]

File: Exiled.Events/Patches/Events/Player/Spawning.cs [hunk 3]
```diff
@@ -27,148 +28,44 @@ namespace Exiled.Events.Patches.Events.Player
...
-
-            LocalBuilder player = generator.DeclareLocal(typeof(Player));
-
-            const int offset = 0;
-            int index = newInstructions.FindLastIndex(instruction => instruction.IsLdarg(1)) + offset;
-
-            newInstructions.InsertRange(
-                index,
-                new[]
+            if (NetworkServer.active && Player.TryGet(hub, out Player player))
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/Spawning.cs:45]:
```suggestion
            if (Player.TryGet(hub, out Player player))
```

File: Exiled.Events/Patches/Events/Player/Spawning.cs [hunk 4]
```diff
@@ -27,148 +28,44 @@ namespace Exiled.Events.Patches.Events.Player
...
+                    if (newRole.ServerSpawnFlags.HasFlag(RoleSpawnFlags.UseSpawnpoint) && fpcRole.SpawnpointHandler != null && fpcRole.SpawnpointHandler.TryGetSpawnpoint(out Vector3 position, out float horizontalRot))
+                    {
+                        oldPosition = position;
+                        fpcRole.FpcModule.MouseLook.CurrentHorizontal = horizontalRot;
+                    }
+
+                    SpawningEventArgs ev = new(player, oldPosition, prevRole);
+
+                    Handlers.Player.OnSpawning(ev);
+                    hub.transform.position = ev.Position;
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/Spawning.cs:59]:
```suggestion

                    hub.transform.position = ev.Position;
```

File: Exiled.Events/Patches/Events/Player/Spawning.cs [hunk 5]
```diff
@@ -27,148 +28,44 @@ namespace Exiled.Events.Patches.Events.Player
...
-                    new(OpCodes.Callvirt, PropertyGetter(typeof(SpawningEventArgs), nameof(SpawningEventArgs.Position))),
-                    new(OpCodes.Callvirt, PropertySetter(typeof(Transform), nameof(Transform.position))),
-
-                    new CodeInstruction(OpCodes.Nop).WithLabels(continueLabel),
-                });
-
-            for (int z = 0; z < newInstructions.Count; z++)
-                yield return newInstructions[z];
+                    SpawningEventArgs ev = new(player, oldPosition, prevRole);
+                    Handlers.Player.OnSpawning(ev);
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/Spawning.cs:64]:
```suggestion
                    Handlers.Player.OnSpawning(new(player, oldPosition, prevRole));
```

File: Exiled.Events/Patches/Fixes/PositionSpawnScp0492Fix.cs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Shared.Rent(instructions);
+
+            Label continueLabel = generator.DefineLabel();
+
+            LocalBuilder player = generator.DeclareLocal(typeof(Player));
+            LocalBuilder eventArgs = generator.DeclareLocal(typeof(SpawningEventArgs));
+
+            const int toRemove = 7;
+
+            int offset = -1;
```
Comment (iopietro) [inline Exiled.Events/Patches/Fixes/PositionSpawnScp0492Fix.cs:44]:
```suggestion
            const int offset = -1;
```

**Review by iopietro — CHANGES_REQUESTED** [review]

File: Exiled.Events/Patches/Events/Player/FlippingCoin.cs (hunk 2 above)
Comment (iopietro) [reply Exiled.Events/Patches/Events/Player/FlippingCoin.cs:46]: Where is it?

File: Exiled.Events/Patches/Events/Player/FlippingCoin.cs [hunk 7]
```diff
@@ -39,32 +43,24 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
-            // Remove both NW events and logic
-            newInstructions.RemoveRange(index, 45);
-
-            // Redirect the old labels to the new starting point
-            newInstructions[index].WithLabels(oldLabels);
+            int index = newInstructions.FindIndex(instruction => instruction.LoadsConstant(109)) + offset;
 
             offset = -1;
-            index = newInstructions.FindLastIndex(instruction => instruction.LoadsField(Field(typeof(Coin), nameof(Coin.RateLimiter)))) + offset;
+            int replaseIndex = newInstructions.FindIndex(instruction => instruction.Calls(PropertyGetter(typeof(Random), nameof(Random.value)))) + offset;
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/FlippingCoin.cs:49]:
```suggestion
            int replaceIndex = newInstructions.FindIndex(instruction => instruction.Calls(PropertyGetter(typeof(Random), nameof(Random.value)))) + offset;
```

File: Exiled.Events/Patches/Events/Player/FlippingCoin.cs [hunk 8]
```diff
@@ -39,32 +43,24 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
...
 
             newInstructions.InsertRange(
                 index,
                 new[]
                 {
                     // Player.Get(ReferenceHub)
-                    new CodeInstruction(OpCodes.Ldloc_0).MoveLabelsFrom(newInstructions[index]),
+                    new CodeInstruction(OpCodes.Ldarg_0).MoveLabelsFrom(newInstructions[index]).MoveLabelsFrom(newInstructions[replaseIndex]),
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/FlippingCoin.cs:56]:
```suggestion
                    new CodeInstruction(OpCodes.Ldarg_0).MoveLabelsFrom(newInstructions[index]).MoveLabelsFrom(newInstructions[replaceIndex]),
```

**Review by iopietro — CHANGES_REQUESTED** [review]

File: Exiled.Events/Patches/Events/Player/FlippingCoin.cs (hunk 7 above)
Comment (iopietro) [inline Exiled.Events/Patches/Events/Player/FlippingCoin.cs:49]: The `switch` at this index doesn't seem to have labels.

**Review by iopietro — CHANGES_REQUESTED** [review]

File: Exiled.API/Features/Roles/Role.cs [hunk 9]
```diff
@@ -184,13 +189,9 @@ protected Role(PlayerRoleBase baseRole)
         /// </summary>
         /// <param name="newRole">The new <see cref="RoleTypeId"/> to be set.</param>
         /// <param name="reason">The <see cref="SpawnReason"/> defining why the player's role was changed.</param>
-        public virtual void Set(RoleTypeId newRole, SpawnReason reason = Enums.SpawnReason.ForceClass)
-        {
-            if (Owner.Role == newRole)
```
[the remark this answers is not in the document]
Comment (IRacle1) [reply Exiled.API/Features/Roles/Role.cs:189]: Yes, exiled patch returns from `InitRole` method if roles are the same, i fixed that

File: Exiled.Events/Patches/Events/Player/FlippingCoin.cs (hunk 7 above)
Comment (IRacle1) [reply Exiled.Events/Patches/Events/Player/FlippingCoin.cs]: It is has
image
image

File: Exiled.API/Features/Roles/Role.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (iopietro) [reply Exiled.API/Features/Roles/Role.cs:189]: Ok.

File: Exiled.Events/Patches/Events/Player/FlippingCoin.cs (hunk 7 above)
Comment (iopietro) [reply Exiled.Events/Patches/Events/Player/FlippingCoin.cs]: Sorry I looked at the operand of br.s and then at the wrong index.

**Review by iopietro — APPROVED** [review]

**Pull request merged**